File: ggalmazor/php/tests/StringCalculatorTest.php
<?php

require_once '../StringCalculatorFactory.php';

/**
 * Screencast de la kata básica (sin DI) en http://vimeo.com/19208035
 */
class StringCalculatorTest extends \PHPUnit_Framework_TestCase {

  protected $stringCalculator;

  public function setUp() {
    $this->stringCalculator = StringCalculatorFactory::create(StringCalculatorFactory::EXTRACTOR_TOKENIZER);
  }

  public function getSumTestCases() {
    return array(
        'Empty string should return zero' => array('str' => "", 'expected' => 0),
        'One number string should return the number' => array('str' => "1", 'expected' => 1),
        'Two number string should return the sum' => array('str' => "1,2", 'expected' => 3),
        'Three number string should return the sum' => array('str' => "1,2,3", 'expected' => 6),
        'New line as separator is accepted and can be mixed with comma' => array('str' => "1\n2,3", 'expected' => 6),
        'Custom one-char separators can be defined' => array('str' => "//;\n1;2", 'expected' => 3),
        'Numbers bigger than 1000 should be ignored' => array('str' => "1,1001", 'expected' => 1),
        'Custom many-char separators can be defined' => array('str' => "//[***]\n1***2***3", 'expected' => 6),
        'Many custom separators can be defined' => array('str' => "//[*][%]\n1*2%3", 'expected' => 6),
    );
  }

  /**
   * @dataProvider getSumTestCases
   */
  public function testSumCases($str, $expected) {
    $this->assertEquals($expected, $this->stringCalculator->add($str));
  }

  public function testNegativeNumbersAreNotAllowed() {
    $this->setExpectedException('Exception');
    $this->stringCalculator->add("1,-2");
  }

}

File: ggalmazor/ruby/StringCalculatorTest.rb
require 'StringCalculator'

describe StringCalculator, "add" do
  
  before(:each) do
    @sc = StringCalculator.new
  end

  sumTestCases = [
    ["returns zero for an empty string", "", 0],
    ["returns the number for a string with a number", "1", 1],
    ["returns the sum for a string with two numbers", "1,2", 3],
    ["returns the sum for a string with three numbers", "1,2,3", 6],
    ["accepts new line as separator", "1\n2,3", 6],
    ["accepts a custom one char separator", "//;\n1;2", 3],
    ["accepts a custom many char separator", "//[***]\n1***2***3", 6],
    ["ignores any number bigger than 1000", "1,1001", 1],
    ["accepts many custom separators", "//[*][%]\n1*2%3", 6]
  ]

  sumTestCases.each{ |testCase|
    it testCase[0] do
      @sc.add(testCase[1]).should == testCase[2]
    end
  }

  it "raises a detailed exception for a string with negative number(s)" do
    lambda{ @sc.add("1,-2") }.should raise_error() { |error|
      error.message.should include("-2")
    }
  end
end
